cep-recipe-serialization.md

```diff
@@ -0,0 +1,1110 @@
+      name: curl
+      version: 8.0.1
+      build_string: h60d57d3_0
+  packaging_settings:
+    archive_type: conda
+    compression_level: 15
+    compression_threads: 1
+```
+
+### Finalized dependencies
+
+The finalized dependencies section contains a "lockfile" of all the dependencies
+exactly as they were used at build-time, for the two environments `build` and
+`host`. It also includes the final `run` and `constrains` specs that are
+computed based on the `build` and `host` dependencies (resolved all
+`run_exports`).
+
+The `finalized_dependencies` section contains the following keys:
+
+```yaml
+finalized_dependencies:
+  build:
+    specs: [list of spec objects]
+    resolved: [list of resolved dependency objects]
+    run_exports: { package_name: run_export } # all run_exports.json that were found among the packages in this environment
+  host:
+    specs: [list of spec objects]
+    resolved: [list of resolved dependency objects]
+    run_exports: { package_name: run_export } # all run_exports.json that were found among the packages in this environment
+  run:
+    depends: [list of dependencies with dependency source information]
+    constrains: [list of constrains with dependency source information]
+    run_exports: optional run_export dictionary
+```
+
+#### Spec objects
+
+A dependency spec contains the `MatchSpec` and additional information about the
+source of the dependency. There are 6 possible `DependencyInfo` versions:
+
+- `compiler`: The dependency comes from a `compiler` expression.
+- `pin_subpackage`: The dependency comes from a `pin_subpackage` expression.
+- `pin_compatible`: The dependency comes from a `pin_compatible` expression.
+- `raw`: The dependency comes from a raw dependency.
+- `run_export`: The dependency comes from a run export.
+- `variant`: The dependency comes from a variant.
+
+Example:
+
+```yaml
+- source: compiler
+  spec: clang_osx-arm64
+- source: raw
+  spec: make >=1.3
+- source: pin_compatible
+  spec: quarto >=1.4.550,<1.5
+- source: pin_subpackage
+  spec: some-pkg-a >=1.0.0,<2
+- source: run_export
+  spec: libzlib >=1.3.1,<1.4.0a0
+  from: host # environment that exported this dependency
+  source_package: zlib # source package that exported this dependency
+- source: variant
+  spec: setuptools 69.*
+  variant: "69" # variant coming from the variant configuration
+```
+
+The list of `resolved` dependencies contains a dictionary with the following
+fields (equivalent to `RepoDataRecord`):
+
+```yaml
+build: hb7217d7_0
+build_number: 0
+depends: []
+license: GPL-2.0-or-later
+license_family: GPL
+md5: fe8efc3385f58f0055e8100b07225055
+name: libtool
+sha256: efe50471d2baea151f2a93f1f99c05553f8c88e3f0065cdc267e1aa2e8c42449
+size: 408403
+subdir: osx-arm64
+timestamp: 1672362003513 # when available
+version: 2.4.7
+fn: libtool-2.4.7-hb7217d7_0.conda
+url: https://conda.anaconda.org/conda-forge/osx-arm64/libtool-2.4.7-hb7217d7_0.conda
+channel: https://conda.anaconda.org/conda-forge/
+```
```

Comment:
Is there a reason to repeat the contents for `RepoDataRecord`? Seems to me that the URL and SHA256 checksum of the package would be enough to recreate the various build environments.

Reply:
yeah, I agree. sha hash and url should be enough. I think it might be interesting to include additional information (e.g. licenses) so that tools could report on what-went-into the package ... Of course, given that we still have access to the packages we could recreate this information from the repodata / packages, too.

Reply:
I think there might be a few more arguments:

- repodata changes over time and this is an accurate snapshot as how it was at the time
- it makes it easier to extract a `conda-lock` file from the rendered recipe
- there might be additioanl opportunities for tooling in the future (license scanning, ...)